include/swift/AST/Evaluator.h

```diff
@@ -416,6 +463,51 @@ class Evaluator {
     return result;
   }
 
+private:
+  template <typename Request,
+            typename std::enable_if<!Request::isDependencySink>::type * = nullptr>
+  void reportEvaluatedResult(const Request &r,
+                             const typename Request::OutputType &o) const {}
+
+  template <typename Request,
+            typename std::enable_if<Request::isDependencySink>::type * = nullptr>
+  void reportEvaluatedResult(const Request &r,
+                             const typename Request::OutputType &o) {
+    r.writeDependencySink(*this, o);
+  }
+
+public:
+  /// Returns \c true if the scope of the current active source cascades.
+  ///
+  /// If there is no active scope, the result always cascades.
+  bool isActiveSourceCascading() const {
+    return getActiveSourceScope() == evaluator::DependencyScope::Cascading;
+  }
+
+  /// Returns the scope of the current active scope.
+  evaluator::DependencyScope getActiveSourceScope() const {
+    if (dependencySources.empty()) {
+      return evaluator::DependencyScope::Cascading;
+    }
+    return dependencySources.back().getInt();
+  }
+
+  /// Returns the active dependency's source file, or \c nullptr if no
+  /// dependency source is active.
```

Comment:
Worthwhile to use `NullPtr` here?

Reply:
Unfortunately, that ties `Evaluator.h` to the definitions of `SourceFile` and `ReferencedNameTracker` because of the template instantiation. It's a change I can make, but it means most every part of the compiler will depend on those headers.